### ISSUE TITLE
Stuttering audio in iOS

### ISSUE THREAD

**Issue by argash** [advplyr/audiobookshelf-app#101]
### Steps to reproduce

1. open ABS iOS app
2. pick audiobook
3. begin playback

### Expected behavior
- Buttery smooth audio

### Actual behavior
- [video of the behavior](https://files.eriksplex.xyz/s/ez4N64M7ProWrkY)
- I think this is only happening with mp3 files. So far I have tested with 2 books that use mp3 and replicated the issue, and one book using mp4 and was not able to replicate. I don't have time at the moment for further testing but if/when I find the time I will update with the results.

### Environment data

Audiobookshelf Version: v1.6.66
Audiobookshelf iOS Beta: v0.9.40-beta

- [ ] Android App?
- [ X] iOS App?

#### iOS Issue
iOS Version: 15.2.1

iPhone model: 13 Pro Max

**Comment by advplyr** [comment]
I remember this happening and I think it was specific to when the audiobook was playing in the background. Can you check if you can pinpoint when the stutter starts happening?

**Comment by argash** [comment]
@advplyr were you able to watch that video before the link expired? I can reshare it if needed.

**Comment by dlahmad** [comment]
I have the same problem. @advplyr it is definitely not specific to the app playing in the background. It happens as soon as I start an audiobook. 

It makes the IOS app basically unusable for me, as a normal playback is not possible. It stops every 1 or 2 seconds.

Device: IPhone XR
IOS: 15.3.1 (Problem also existent with previous versions)

**Comment by advplyr** [comment]
@rasmuslos has done a lot of work on the ios app over the last week that may resolve this issue.

We are in the middle of a big rewrite of the data model so I can't really give an ETA on when that will be published. I'll be posting updates regularly in the discord server though.

**Comment by rasmuslos** [comment]
I am not entirely sure, but when the server transcodes on the fly it is possible that it isn't fast enough to keep up with your progress

**Comment by advplyr** [comment]
I didn't think of that. For android I configured the buffer time, maybe the iOS player just needs to be configured properly.

**Comment by advplyr** [comment]
> > I didn't think of that. For android I configured the buffer time, maybe the iOS player just needs to be configured properly.
> 
> I mean i could set a buffer size, but normally iOS does this automatically https://developer.apple.com/documentation/avfoundation/avplayeritem/1643630-preferredforwardbufferduration?preferredLanguage=occ

Android does it automatically too, I just increased the buffer size.
In android and HLS.js they have settings to set how long to wait before requesting a segment again after it fails. Maybe we need to find that setting so the audio player gives the server some more time to transcode.

**Comment by rasmuslos** [comment]
Maybe it is related to this ffmpeg warning
```
Cannot read BOM value, input too short
Error reading frame UFID, skipped
```

**Comment by advplyr** [comment]
Please test the latest iOS release, your audio files should direct play now.

**Comment by seefernando** [comment]
I just downloaded the iOS app (0.9.47-Beta and I'm seeing the same behaviour.  Seems to only affect audiobooks that are one single file (only tested with 3 .mp3 files) but works file with .m4b files.

the MP3s play fine in firefox on windows and android, but also stutter when played in firefox on iOS.

**Comment by advplyr** [comment]
@seefernando I believe this is related to the browser not being able to direct play the audio files, so the server is transcoding the audio files in a streamable format (HLS).

Since the transcode is happening on the fly we are reporting back to the client that it is finished when it actually isn't so that the client can start playing right away.

The iOS audio player doesn't seem to handle this well and is trying to request segments that aren't yet created causing the stutter effect. Unfortunately the iOS audio player has very little documentation or just doesn't support any customization with how segments are requested (or I just haven't found it yet).

There must be a way to handle on-the-fly HLS transcoding eloquently since it seems other services handle this, but I'm still in the beginning of learning iOS development and hoping some iOS dev helps us out!

Fyi: Android HLS player (ExoPlayer) and the web HLS.js both allow customizing what happens when segments fail, how many segments ahead to request, etc. so we can handle it smoothly there.

**Comment by seefernando** [comment]
@advplyr thanks, its unfortunate for the lack of control.  As a test I wanted to convert an MP3 to HLS (is that even a valid idea?), and to transfer the files I used nextcloud.  Interestingly, from the nextcloud web ui (in firefox) I was able to stream the mp3 file without any stuttering.

I'm still looking for a way to convert from MP3 to HLS, will report back if I find anything.  Though its interesting that the Nextcloud UI is able to play the file through browser while ABS stutters through the same browser.
